source/agora/test/NetworkDiscovery.d

```diff
@@ -83,3 +87,65 @@ unittest
                format("Node %s has %d peers: %s", key, addresses.length, addresses));
     }
 }
+
+///
+unittest
+{
+    import core.thread;
+    import core.atomic;
+
+    __gshared int call_count = 0;
+    __gshared NodePair impersonator_node;
+
+    static class ImpersonatorValidator : TestValidatorNode
+    {
+        mixin ForwardCtor!();
+
+        /// GET /public_key
+        protected override Identity getPublicKey (PublicKey key) nothrow @safe
+        {
+            () @trusted { call_count++; } ();
+            return Identity(this.config.validator.key_pair.address);
+        }
+    }
+
+    static class ImpersonatorManager : TestAPIManager
+    {
+        mixin ForwardCtor!();
+
+        public override void createNewNode (Config conf, string file = __FILE__,
+            int line = __LINE__)
+        {
+            if (impersonator_node != NodePair.init && conf.validator.enabled)
+                this.addNewNode!TestValidatorNode(conf, file, line);
+            else if (conf.validator.enabled)
+            {
+                this.addNewNode!ImpersonatorValidator(conf, file, line);
+                impersonator_node = this.nodes[$ - 1];
+            }
+            else
+                this.addNewNode!TestFullNode(conf, file, line);
+        }
+    }
+
+    TestConf conf;
+    auto network = makeTestNetwork!ImpersonatorManager(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+
+    int count = 0;
+    while (call_count < GenesisValidators - 1)
+    {
+        if (count++ > 30) // 3 secs
+            break;
+        Thread.sleep(100.msecs);
+    }
```

Comment:
Aren't you just sleeping 3s here ?

Reply:
How? `call_count` should eventually reach `GenesisValidators - 1` and terminate the loop.

Reply:
Brain was disconnected, ignore.